# popRR - recombination estimation from pooled genotyping and sequencing

This tool performs **genetic map** and **recombination rate** estimation from pooled population samples.
Highly similarities regarding the recombination rate were oberserved comparing this pool based estimation and Haldane derived mappings


![Requirements](https://smct-management.de/wp-content/uploads/2020/04/Minium-Requirements-IATF-16949-1024x384.jpeg.webp)

## Requirements / experimental set up 

- a crossing population beyond F2
- the sampling of a represetive number of genotypes for each population ( e.g., in a population of 10,000 siblings, 200 to 500 genotypes should be selected for testing)
- a **reference genome**
- whole genome pool sequencing data with a decend coverage per locus (50x are advised; see https://github.com/mischn-dev/HAFcall if you are planing to reduce sequencing levels to 5 or 10x)
    - one pool per population (or you create additional replications - we advise you to used different genotypes for each replicate created)



##
##
##


## Work flow 

Perform **trimming** of your sequence data if required and progress with the **alignment**. 
For complex plant genomes like, maize, barley or wheat, we observed *bwa* and *bowtie2* to perform best.

**Filter your** aligned reads throughly. The recombination rate estimations depend on this filtering step, as following errors might reduce the accuracy:

- secondary alignments
- supplementary alignments
- low quality alignments
- duplicated fragments

Contrasting to typical variant calling of homogygote genotypes, we need to make sure the ratio of *reference* to *alternative* base calls is as unbiased as possible.

We will generally expect `REF/ALT` variant calls of `20/30` or `10/40`, contrasting to `2/48` - what we would expect for an homozygote single genotyping.

So it is very crucial to maintain high quality from the alignment and the variant calling, as biases introduced by the alignment might directly impact the accuracy of the recombinate rate estimation.

You can filter your aligned bam file by applying 

```
# X = number of threads to use

# filter the alignments - this one works for the alignment flags generated by bwa 
sambamba_v0.8.2 view -t X -h -f bam -p -F "mapping_quality >= 30 and not (unmapped or secondary_alignment or 
supplementary) and not ([XA] != null or [SA] != null) and proper_pair" -o filtered.bam ALIGNED.bam

# remove the duplicates and sort
sambamba_v0.8.2 markdup -r -t X -p filtered.bam sambamba_filtered_markdup.bam

# sort the file by position again - important to do for variant calling
sambamba_v0.8.2 sort -t X -p -o filtered_sorted.bam sambamba_filtered_markdup.bam
```
or using your own script. We observed *sambamba* to be very fast and accurate, so we suggest using this package <https://lomereiter.github.io/sambamba/>

Find further infos for the filtering by sambamba here <https://github.com/biod/sambamba/wiki/%5Bsambamba-view%5D-Filter-expression-syntax>

*Otherwise, you can also use **samtools** to filter your alignments by using this pipeline:* 

```
# X = numer of threads to use

# filter the alignments
samtools view -F 2304 -q 30 -f 1 -@ X -o filtered.bam ALIGNED.bam 
# remove duplicates
samtools markdup -@ X -r filtered.bam filtered_markdup.bam 
# sort by position 
samtools sort -@ X -o filtered_sorted.bam filtered_markdup.bam 
```

What does -F 2034 mean? - have a look here <https://broadinstitute.github.io/picard/explain-flags.html>


### **Variant calling**

If possible, call the variants of *all* pool samples in one run.

The recombination rate estimation for each variant locus depends on the **allelic depth** tag *AD*, so ensure your chosen variant caller supports this output format. *Bcftools* is the option used in the procudre of validating popRR.

The code to generate the variant calls with *bcftools* is:

```
# define the input variables
REF="/RefGenome.fa" 
mydir="StoreFolder" # e.g. ~/Documents

bcftools mpileup -Ov -q 25 -Q 30 -a FORMAT/AD -I -f $REF $mydir/*filtered_sorted.bam | bcftools call -vm -Ov > $mydir/variants.vcf
```

**Additional hints for the SNP calling:**
- if available, use an *SNP reference map*, like <http://ftp.ensemblgenomes.org/pub/plants/release-52/variation/vcf/hordeum_vulgare/>
    - this is helpful to avoid false positive detections and will improve the overall recombination rate accuracy
- exclude Indels, as these tend to have higher false positive rates



## Variant filtering and data preparaption


To fit into the desired data format and filter out low quality SNPs, the uncompressed variant calling format (*vcf*) files need to be processed. 

You can use either `bcftools` or `vcftools` (<http://vcftools.sourceforge.net/>) to complete the task.
Use on of either functions:
```
# the vcftools way 
vcftools --vcf variants.vcf --out outfFileName --max-alleles 2 --min-alleles 2 --minQ 40 --extract-FORMAT-info AD

# the bcftools way
bcftools view -m2 -M2 -v snps -o variants_biallele.vcf variants.vcf
bcftools query -i 'QUAL>40' -f '%CHROM\t%POS\t[%AD\t]\n' -o outFileName.vcf variants_biallele.vcf
```


Besides the *Position* (Chromosome and pysical position on the reference genome), the *allelic depth AD* is required for each sample.

The final table should look like:
| CHROM | POS  | Sample1  | Sample2 | Sample3 |
| ------- | --- | --- | --- | --- |
chr2H |	268269	|127,98 | 	175,128	|187,119 |
chr2H |	352795	| 133,79	| 190,173	| 214,159 |
chr3H |	556914877 | 116,111 |	133,178 |	159,157 |
chr5 |	582298638 | 120,115	| 157,145 |	147,153|


The names of your *samples* can also look like `/Data/recrate/sample1` instead of `Sample1` - both are fine



## Perform the **recombination rate estimation** and the **genetic map** estimation using popRR functions

You as *End User* have the choice to run it either in `julia` or `R`. The result is the same, but the computational time is quite drastically decreased in the `julia` environment (about 20 to 100 times faster)

So our recommendations is to use the julia function package - save some energy and resources by smart programming :)

![sqirrel](https://naturschutz.ch/wp-content/uploads/2018/10/cropped-Eichh%C3%B6rnchen--1068x580.jpg)

